### ISSUE TITLE
S3: "Failed to set modification time: InvalidRange"

### ISSUE THREAD

**Issue by Platzii** [rclone/rclone#3778]
#### What is the problem you are having with rclone?

Bug in the latest version of rclone (`v1.50.2`) where changing the modified date of a file fails on a S3 remote.
In version `v1.49.5` this bug was not present yet.

#### What is your rclone version (output from `rclone version`)

`v1.50.2`
(also tested with the latest beta: `v1.50.2-073-ge2773b3b`)

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)

Ubuntu 16.04.4 LTS, 64-bit

####  Which cloud storage system are you using? (eg Google Drive)

S3-compatible interface (own hosted Ceph cluster with RADOS GW)

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)

`./rclone-v1.50.2-linux-amd64/rclone copy --s3-chunk-size=50M /home/user/myfile.dat ceph-custer:data/mydata`

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)

v1.50.2 (fail):
```
root@myhost ~ # ./rclone-v1.50.2-linux-amd64/rclone -vvv copy --s3-chunk-size=50M /home/user/myfile.dat ceph-custer:data/mydata
2019/12/02 13:25:57 DEBUG : rclone: Version "v1.50.2" starting with parameters ["./rclone-v1.50.2-linux-amd64/rclone" "-vvv" "copy" "--s3-chunk-size=50M" "/home/user/myfile.dat" "ceph-custer:data/mydata"]
2019/12/02 13:25:57 DEBUG : Using config file from "/root/rclone-v1.50.2-linux-amd64/rclone.conf"
2019/12/02 13:25:57 DEBUG : myfile.dat: Modification times differ by -18.718573627s: 2019-12-02 11:36:58.718573627 +0100 CET, 2019-12-02 11:36:40 +0100 CET
2019/12/02 13:25:58 DEBUG : myfile.dat: MD5 = 45cd09bdd16261f03588c31630decefd OK
2019/12/02 13:26:03 ERROR : myfile.dat: Failed to set modification time: InvalidRange:
	status code: 416, request id: tx00000000000005bed0aa5-005de502da-97f07-default, host id:
2019/12/02 13:26:03 DEBUG : myfile.dat: Unchanged skipping
2019/12/02 13:26:03 ERROR : Attempt 1/3 failed with 1 errors and: InvalidRange:
	status code: 416, request id: tx00000000000005bed0aa5-005de502da-97f07-default, host id:
2019/12/02 13:26:03 DEBUG : myfile.dat: Modification times differ by -18.718573627s: 2019-12-02 11:36:58.718573627 +0100 CET, 2019-12-02 11:36:40 +0100 CET
2019/12/02 13:26:04 DEBUG : myfile.dat: MD5 = 45cd09bdd16261f03588c31630decefd OK
2019/12/02 13:26:08 ERROR : myfile.dat: Failed to set modification time: InvalidRange:
	status code: 416, request id: tx00000000000005bed0acf-005de502e0-97f07-default, host id:
2019/12/02 13:26:08 DEBUG : myfile.dat: Unchanged skipping
2019/12/02 13:26:08 ERROR : Attempt 2/3 failed with 1 errors and: InvalidRange:
	status code: 416, request id: tx00000000000005bed0acf-005de502e0-97f07-default, host id:
2019/12/02 13:26:08 DEBUG : myfile.dat: Modification times differ by -18.718573627s: 2019-12-02 11:36:58.718573627 +0100 CET, 2019-12-02 11:36:40 +0100 CET
2019/12/02 13:26:09 DEBUG : myfile.dat: MD5 = 45cd09bdd16261f03588c31630decefd OK
2019/12/02 13:26:14 ERROR : myfile.dat: Failed to set modification time: InvalidRange:
	status code: 416, request id: tx00000000000005bed0af0-005de502e5-97f07-default, host id:
2019/12/02 13:26:14 DEBUG : myfile.dat: Unchanged skipping
2019/12/02 13:26:14 ERROR : Attempt 3/3 failed with 1 errors and: InvalidRange:
	status code: 416, request id: tx00000000000005bed0af0-005de502e5-97f07-default, host id:
2019/12/02 13:26:14 Failed to copy: InvalidRange:
	status code: 416, request id: tx00000000000005bed0af0-005de502e5-97f07-default, host id:
root@myhost ~ # echo $?
1
```

v1.49.5 (success)
```
root@myhost ~ # ./rclone-v1.49.5-linux-amd64/rclone -vvv copy --s3-chunk-size=50M /home/user/myfile.dat ceph-custer:data/mydata
2019/12/02 13:26:23 DEBUG : rclone: Version "v1.49.5" starting with parameters ["./rclone-v1.49.5-linux-amd64/rclone" "-vvv" "copy" "--s3-chunk-size=50M" "/home/user/myfile.dat" "ceph-custer:data/mydata"]
2019/12/02 13:26:23 DEBUG : Using config file from "/root/rclone-v1.49.5-linux-amd64/rclone.conf"
2019/12/02 13:26:23 DEBUG : myfile.dat: Modification times differ by -18.718573627s: 2019-12-02 11:36:58.718573627 +0100 CET, 2019-12-02 11:36:40 +0100 CET
2019/12/02 13:26:24 DEBUG : myfile.dat: MD5 = 45cd09bdd16261f03588c31630decefd OK
2019/12/02 13:26:24 INFO  : myfile.dat: Updated modification time in destination
2019/12/02 13:26:24 DEBUG : myfile.dat: Unchanged skipping
2019/12/02 13:26:24 INFO  :
Transferred:   	         0 / 0 Bytes, -, 0 Bytes/s, ETA -
Errors:                 0
Checks:                 1 / 1, 100%
Transferred:            0 / 0, -
Elapsed time:          0s

2019/12/02 13:26:24 DEBUG : 5 go routines active
2019/12/02 13:26:24 DEBUG : rclone: Version "v1.49.5" finishing with parameters ["./rclone-v1.49.5-linux-amd64/rclone" "-vvv" "copy" "--s3-chunk-size=50M" "/home/user/myfile.dat" "ceph-custer:data/mydata"]
root@myhost ~ # echo $?
0
```

**Comment by ncw** [comment]
Can you do something like

    rclone touch ceph-custer:data/mydata/myfile.dat -vv --dump bodies --low-level-retries 1 --retries 1

and post the output for the working rclone and the non working one?

When I try this with s3 and rclone 1.49.5 I get these two requests which look very similar (they both work).

```
2019/12/02 15:01:27 DEBUG : HTTP REQUEST (req 0xc000481100)
2019/12/02 15:01:27 DEBUG : PUT /rclone/file.txt HTTP/1.1
Host: s3.eu-west-2.amazonaws.com
User-Agent: rclone/v1.49.5
Content-Length: 0
Authorization: XXXX
Content-Type: text/plain; charset=utf-8
X-Amz-Acl: private
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Copy-Source: rclone/file.txt
X-Amz-Date: 20191202T150127Z
X-Amz-Meta-Mtime: 1575298887.567934566
X-Amz-Metadata-Directive: REPLACE
Accept-Encoding: gzip
```

and

latest beta:

```
2019/12/02 15:03:34 DEBUG : HTTP REQUEST (req 0xc00049b100)
2019/12/02 15:03:34 DEBUG : PUT /rclone/file.txt HTTP/1.1
Host: s3.eu-west-2.amazonaws.com
User-Agent: rclone/v1.50.2-073-g56c0953e-fix-1205-order-by-beta
Content-Length: 0
Authorization: XXXX
Content-Type: text/plain; charset=utf-8
X-Amz-Acl: private
X-Amz-Content-Sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
X-Amz-Copy-Source: rclone/file.txt
X-Amz-Date: 20191202T150334Z
X-Amz-Meta-Mtime: 1575299014.473292137
X-Amz-Metadata-Directive: REPLACE
Accept-Encoding: gzip
```

Is your file very large maybe?

**Comment by Platzii** [comment]
Hi @ncw, thank you for your reply.
The file is 287M large, not that big imho.

I've uploaded the output you requested over here: https://gist.github.com/Platzii/47224e18ae5dc7c364b49f13cb47e81b

**Comment by ncw** [comment]
I see what is happening...

It is the multipart copy code that has gone wrong.

It turns out there is an off by one bug which is the cause of the error, but it really shouldn't be doing multipart copies for such small files.  I've attempted for fix both issues here - can you have a go?

https://beta.rclone.org/branch/v1.50.2-074-g8b19122f-fix-3778-s3-copy-beta/ (uploaded in 15-30 mins)

**Comment by Platzii** [comment]
Looks like you got it right, both the touch- and copy-command appear to be working with the beta version. Thanks!

**Comment by ncw** [comment]
Thanks for testing

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.51
